internal/node/npm_package_bin.bzl

```diff
@@ -21,20 +21,31 @@ _ATTRS = {
 # because the output_dir is a tree artifact
 # so we weren't able to give it a label
 def _expand_location(ctx, s):
-    outdir_segments = [ctx.bin_dir.path, ctx.label.package]
+    rule_dir = [ctx.bin_dir.path, ctx.label.package]
+
     if ctx.attr.output_dir:
-        # We'll write into a newly created directory named after the rule
-        outdir_segments.append(ctx.attr.name)
+        if s.find("$@") != -1:
+            fail("""$@ substitution may only be used with output_dir=False.
+            Upgrading rules_nodejs? Maybe you need to switch from $@ to $(@D)
+            See https://github.com/bazelbuild/rules_nodejs/releases/tag/0.42.0""")
 
-    if not ctx.attr.output_dir:
+        # We'll write into a newly created directory named after the rule
+        output_dir = [ctx.bin_dir.path, ctx.label.package, ctx.attr.name]
+    else:
         if s.find("$@") != -1 and len(ctx.outputs.outs) > 1:
             fail("""$@ substitution may only be used with a single out
             Upgrading rules_nodejs? Maybe you need to switch from $@ to $(RULEDIR)
             See https://github.com/bazelbuild/rules_nodejs/releases/tag/0.42.0""")
         s = s.replace("$@", ctx.outputs.outs[0].path)
+        if len(ctx.outputs.outs) == 1:
+            output_dir = ctx.outputs.outs[0].dirname.split("/")
+        else:
+            output_dir = rule_dir[:]
```

Comment:
nit: not sure why you need to make a copy here

Reply:
No good reason looks like 🤷‍♂   I think i had an `output_dir.append(something)` at one point